python/dgl/nn/pytorch/softmax.py

```diff
@@ -0,0 +1,104 @@
+"""Torch modules for graph related softmax."""
+# pylint: disable= no-member, arguments-differ
+import torch as th
+from torch import nn
+
+from ... import function as fn
+
+__all__ = ['EdgeSoftmax']
+
+class EdgeSoftmax(nn.Module):
+    r"""Apply softmax over signals of incoming edges.
+
+    For a node :math:`i`, edgesoftmax is an operation of computing
+
+    .. math::
+      a_{ij} = \frac{\exp(z_{ij})}{\sum_{j\in\mathcal{N}(i)}\exp(z_{ij})}
+
+    where :math:`z_{ij}` is a signal of edge :math:`j\rightarrow i`, also
+    called logits in the context of softmax. :math:`\mathcal{N}(i)` is
+    the set of nodes that have an edge to :math:`i`.
+
+    An example of using edgesoftmax is in
+    `Graph Attention Network <https://arxiv.org/pdf/1710.10903.pdf>`__ where
+    the attention weights are computed with such an edgesoftmax operation.
+
+    Parameters
+    ----------
+    logits_name : str
+        A field name such that g.edata[logits_name] gives the logits :math:`z_{ij}`.
+        based on which edgesoftmax will be computed.
+    max_logits_name : str
+        A field name such that g.ndata[max_logits_name] will store
+        :math:`\max_{j\in\mathcal{N}(i)}z_{ij}`.
+    normalizer_name : str
+        A field name such that g.ndata[normalizer_name] will store
+        :math:`\sum_{j\in\mathcal{N}(i)}exp(z_{ij})`
+    """
+    def __init__(self,
+                 logits_name,
+                 max_logits_name,
+                 normalizer_name):
```

Comment:
Give default names for them?

Reply:
Though the chance is not very high, I want to avoid any potential risk of undesired overwriting.

Reply:
We could also detect whether the name has been taken and use another one. Then those temporary variables will not be seen by the user.

Reply:
Done.

Reply:
Then, it's totally safe to remove those "_name" args from the arglist. To think it as a normal pytorch module, the internal local variables are not exposed at all. So are these names. We could also change `GraphConv` in the same way. What do you think?

Reply:
Yeah, I agree. I just forgot to make those changes. Thanks.